File: S2_KP/Lectures/codes02/SuperTest.cpp
#include<iostream>

using namespace std;

class Baza {
   public:
  int a;
  Baza() { a=0; cout << "Domyslny konstruktor klasy Baza" << endl; }
  Baza(int i) { a=i; cout << "Konstruktor klasy Baza" << endl; }
  virtual ~Baza() { cout << "Destruktor klasy Baza" << endl; }
  virtual void m1() { cout << "Baza.m1()" << endl; }
  virtual void m2() { cout << "Baza.m2()" << endl; }
};

class Potomna : public Baza {
  public:
  // Wywolywanie konstruktora bazowego
  Potomna() : Baza(0) { cout << "Domyslny konstruktor klasy Potomna" << endl; }
  Potomna(int i) { a=i; cout << "Konstruktor klasy Potomna" << endl; }
  ~Potomna() { cout << "Destruktor klasy Potomna" << endl; }
  void m3() { cout << "Potomna.m3()" << endl; Baza::m2(); }
  void m2() { cout << "Potomna.m2()" << endl; }
};

int main(int argc, char* argv[]) {

    cout << "new a ------" << endl;
    Baza* a = new Potomna();
    cout << "new b ------" << endl;
    Potomna* b = new Potomna(1);
    cout << "new c ------" << endl;;
    void* c = new Baza();

	// m3 nie jest metoda klasy Baza
//    a->m3(); // błąd kompilatora

	// Niepoprawnie. 
	// Potomna* d = new Baza();

    cout << "b.m1() ------" << endl;
    b->m1();

	// m2 wywola sie z klasy Potomna, bo obiekt a jest tworzony na podstawie Potomna. Ale przy 
	// zalozeniu ze m2 jest zdefiniowana jako virtual
    cout << "a.m2() ------" << endl;
    a->m2();

	// Poprawnie, bo a jest typu Potomna
    ((Potomna*)a)->m2();

    cout << "b.m3() ------" << endl;
    b->m3();

    cout << "a ------" << endl;
    delete a;
    cout << "b ------" << endl;
    delete b; 
    cout << "c ------" << endl;
    delete (Baza*)c;

	cout << "new a1 ------" << endl;
	Baza a1;

}

File: S2_KP/Lectures/codes02/SuperTest.java
class Baza {
  	int a; 
	Baza() { a=0; System.out.println("Domyslny konstruktor klasy Baza"); }  
	Baza(int i) { a=i; System.out.println("Konstruktor klasy Baza" ); } 
	public void m1() { System.out.println("Baza.m1()"); }  
	public void m2() { System.out.println("Baza.m2()"); }
}


class Potomna extends Baza {
	Potomna() { super(0); System.out.println("Domyslny konstruktor klasy Potomna"); }  
	Potomna(int i) { a=i; System.out.println("Konstruktor klasy Potomna" ); }  
	public void m3() { System.out.println("Potomna.m3()"); super.m2(); } 
	public void m2() { System.out.println("Potomna.m2()"); }
}

	
public class SuperTest {
	public static void main(String[] args) {   


            System.out.println("new a ------");         
            // Poprawnie. Klasa Potomna dziedziczy po klasie Baza
            Baza a = new Potomna();   

            System.out.println("new b ------");   
            Potomna b = new Potomna(1);  
            
            System.out.println("new c ------");              
            // Poprawnie. Kazda klasa dziedziczy po klasie Object
            Object c = new Baza();

            
            // Niepoprawnie. Nie kazda klasa bazowa jest potomna. 
            // Potomna p = new Baza();
            
            // Niepoprawnie. m3 jest zdefiniwane z klasie Potomna, a nie Baza.
            // a.m3();   
            
            System.out.println("b.m1() ------");  
            b.m1();  
    
            System.out.println("a.m2() ------"); 
            // Wywola sie metoda z klasy Potomna, chociaz jest dla obiektu Baza
            a.m2();    
            System.out.println("b.m3() ------");  
            b.m3();   
		
            // Sprawdza czy a jest typu Potomna
            if( a instanceof Potomna ) {     
		System.out.println("a wskazuje na obiekt klasy Potomna");    
		((Potomna) a).m3();
            }
    
            if( c instanceof Potomna )      
		System.out.println("c wskazuje na obiekt klasy Potomna");  
            }
}
